haystack_experimental/utils/jinja_chat_extension.py

```diff
@@ -0,0 +1,203 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+from typing import Any, Callable, List, Optional, Union, get_args
+
+from haystack import logging
+from jinja2 import TemplateSyntaxError, nodes
+from jinja2.ext import Extension
+
+from haystack_experimental.dataclasses.chat_message import (
+    ChatMessage,
+    ChatMessageContentT,
+    ChatRole,
+    TextContent,
+    _deserialize_content_part,
+    _serialize_content_part,
+)
+
+logger = logging.getLogger(__name__)
+
+START_TAG = "<haystack_content_part>"
+END_TAG = "</haystack_content_part>"
+
+class ChatMessageExtension(Extension):
+    """
+    A Jinja2 extension for creating structured chat messages with mixed content types.
+
+    This extension provides a custom {% message %} tag that allows creating chat messages
+    with different attributes (role, name, meta) and mixed content types (text, images, etc.).
+
+    Inspired by [Banks](https://github.com/masci/banks).
+
+    Example:
+    {% message role="system" %}
+    You are a helpful assistant. You like to talk with {{user_name}}.
+    {% endmessage %}
+
+    {% message role="user" %}
+    Hello! I am {{user_name}}. Please describe the images.
+    {% for image in images %}
+    {{ image | for_template }}
+    {% endfor %}
+    {% endmessage %}
+
+    ### How it works
+    1. The {% message %} tag is used to define a chat message.
+    2. The message can contain text and other structured content parts.
+    3. To include a structured content part in the message, the `| for_template` filter is used.
+       The filter serializes the content part into a JSON string and wraps it in a `<haystack_content_part>` tag.
+    4. The `_build_chat_message_json` method of the extension parses the message content parts,
+       converts them into a ChatMessage object and serializes it to a JSON string.
+    5. The obtained JSON string is usable in the ChatPromptBuilder component, where templates are rendered to actual
```

Comment:
I'm interested to see step 5! Since it feels a little surprising to me that you need to make a ChatMessage then JSONify it and then recreate it again in ChatPromptBuilder. But I'm sure there is a reason for it :D

Reply:
I agree that it is a bit tricky. The main reason is that Jinja2 mostly work with strings: this is especially true for the [`SandboxedEnvironment`](https://jinja.palletsprojects.com/en/stable/sandbox/) that we use for security reasons.

This forces us to use JSON strings as an intermediate format.

During my investigation for POC, I found that bypassing this limitation would require using `NativeEnvironment`, which introduces security risks.

Reply:
Okay no worries then! Thanks for the explanation.